pkg/volume/plugins.go

```diff
@@ -746,11 +746,10 @@ func (pm *VolumePluginMgr) logDeprecation(plugin string) {
 // If it is, initialize all probed plugins and replace the cache with them.
 func (pm *VolumePluginMgr) refreshProbedPlugins() {
 	events, err := pm.prober.Probe()
-	if err != nil {
-		klog.ErrorS(err, "Error dynamically probing plugins")
-		return // Use cached plugins upon failure.
-	}
 
+	// because the probe function can return a list of valid plugins
+	// even when an error is present we still must add the plugins
+	// or they will be skipped because each event only fires once
```

Comment:
This breaks the prober api, however:

```
type DynamicPluginProber interface {
	Init() error

	// If an error occurs, events are undefined.
	Probe() (events []ProbeEvent, err error)
}
```

Maybe the current implementation happens to work and return valid events, at least on the error you see, but there's no guarantee a different error would't return undefined events that would cause bad behavior when iterating through them.

Also, flexvolumes are deprecated in 1.23 and I don't know what the plan for 1.24 is going to be---are we tearing out the code? @xing-yang 

/cc @xing-yang

Reply:
Right, my code is relying on an implementation specific detail of flexVolumeProber. I suppose I should update the flexVolumeProber to return failed events instead of an error aggregate when individual plugins fail.

Reply:
On second thought if the plugin is skipped then no error should be generated and no event either.

Reply:
Maybe ping Xing on slack (xyang on kubernetes slack) and see if she has more of an opinion around the flexvolume stuff.

Reply:
Looks like the implementation of flexVolumeProber is not consistent with the definition of the Prober api.

Since FlexVolume is deprecated in 1.23, I think non-critical bugs should not be fixed any more. Looks like this problem has a workaround. We don't have an immediate plan to remove the FlexVolume code as there are still lots of people using it.

Reply:
After the pods are stuck in pending due to failed mounts because the plugins are not yet registered, restarting kubelet will fix the issue. The issue with this work around is that you first have to figure out what's causing the mount failure and then you have make sure that all other plugins are present before kubelet restart. Sometimes pods will dynamically add flex driver volume plugins an example of this is calico, which is how I originally found this issue.

The prober api does not allow to my knowledge a way to associate a specific error to a specific plugin while allowing other plugins to be successfully added. Currently any detected failure will mean no volume plugins can be installed until the broken plugin is fixed. Even a restart will not fix this issue.

Reply:
@jsafrane Can you take a look

Reply:
/cc @jsafrane

Reply:
I don't think this has a good workaround - a faulty flex plugin can break registration of all other flex plugins on the same node. Kubelet restart may not help. Question is, how to fix it nicely.

`Prober` interface is used only in tests and here to collect flex volume drivers.
`Probe()` is really simple and it calls either `probeMap` or `probeAll`. They both check **all** flex drivers (either collected from inotify events or read from `/usr/libexec/kubernetes/kubelet-plugins/volume/exec/`) and collects **all** errors.

[`probeAll` can also return an error when the initial `ReadDir` of the directory with drivers fails, but that's IMO fine, it will return empty list of drivers].

IMO it's OK to change description of `Probe()` that it collects as many drivers as it can and returns errors for the rest. Leave the rest of this PR as it is.

Reply:
BTW, would it be possible to add an unit test to ensure `Probe()` behaves as described above?

Reply:
@jsafrane I will update the `Probe()` comment to reflect the functionality. I will add or update a test in the `flexVolumeProber` tests to ensure that it behaves as described above.

Reply:
@jsafrane added a unit test and updated the api comment